### PR TITLE
IPAM Host Resource and Data source Implementation

### PR DESCRIPTION
**Pull request by unasra**

### CODE REVIEW

File: internal/provider/provider.go [hunk 1]
```diff
@@ -90,6 +90,7 @@ func (p *BloxOneProvider) Resources(ctx context.Context) []func() resource.Resou
 func (p *BloxOneProvider) DataSources(ctx context.Context) []func() datasource.DataSource {
 	return []func() datasource.DataSource{
 		ipam.NewIpSpaceDataSource,
+		ipam.NewIpamHostDataSource,
```
Comment (mathewab) [inline internal/provider/provider.go:93]: I don't see an entry in `Resources` function for ipam host

File: internal/service/ipam/model_ipamsvc_host_name.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+func (m *IpamsvcHostNameModel) Expand(ctx context.Context, diags *diag.Diagnostics) *ipam.IpamsvcHostName {
+	if m == nil {
+		return nil
+	}
+	to := &ipam.IpamsvcHostName{
+		Alias:       m.Alias.ValueBoolPointer(),
+		Name:        m.Name.ValueString(),
+		PrimaryName: m.PrimaryName.ValueBoolPointer(),
+		Zone:        m.Zone.ValueString(),
+	}
```
Comment (mathewab) [inline internal/service/ipam/model_ipamsvc_host_name.go:71]: Same comment. 
Can you replace with `flex.Expand...`

File: internal/service/ipam/model_ipamsvc_ipam_host.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+func (m *IpamsvcIpamHostModel) Expand(ctx context.Context, diags *diag.Diagnostics) *ipam.IpamsvcIpamHost {
+	if m == nil {
+		return nil
+	}
+	to := &ipam.IpamsvcIpamHost{
+		Addresses:           flex.ExpandFrameworkListNestedBlock(ctx, m.Addresses, diags, ExpandIpamsvcHostAddress),
+		AutoGenerateRecords: m.AutoGenerateRecords.ValueBoolPointer(),
+		Comment:             m.Comment.ValueStringPointer(),
```
Comment (mathewab) [inline internal/service/ipam/model_ipamsvc_ipam_host.go:115]: can you use `flex.Expand...` instead?

File: internal/service/ipam/model_ipamsvc_ipam_host.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	"addresses": schema.ListNestedAttribute{
+		NestedObject: schema.NestedAttributeObject{
+			Attributes: IpamsvcHostAddressResourceSchemaAttributes,
+		},
+		Optional:            true,
+		MarkdownDescription: `The list of all addresses associated with the IPAM host, which may be in different IP spaces.`,
+	},
+	"auto_generate_records": schema.BoolAttribute{
+		Optional:            true,
+		Computed:            true,
```
[the remark this answers is not in the document]
Comment (venkat-iblox) [reply internal/service/ipam/model_ipamsvc_ipam_host.go:54]: @mathewab Isn't model auto generated?

[the remark this answers is not in the document]
Comment (mathewab) [reply internal/service/ipam/model_ipamsvc_ipam_host.go:54]: It is, but the auto generated model can't be used as it is. You most certainly need changes to be made for default values, validators, plan modifiers and in some cases even in the FlEx logic

File: internal/provider/provider.go (hunk 1 above)
Comment (unasra) [reply internal/provider/provider.go:94]: Added

File: internal/service/dns_config/.openapi-generator-ignore
[the remark this answers is not in the document]
Comment (unasra) [reply internal/service/dns_config/.openapi-generator-ignore:1]: All files are deleted

File: internal/service/ipam/api_ipam_host_data_source.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	resp.Schema = schema.Schema{
+		MarkdownDescription: ``,
+		Attributes: map[string]schema.Attribute{
+			"filters": schema.MapAttribute{
+				Description: "Filter are used to return a more specific list of results. Filters can be used to match resources by specific attributes, e.g. name. If you specify multiple filters, the results returned will have only resources that match all the specified filters.",
+				ElementType: types.StringType,
+				Optional:    true,
+			},
+			"tag_filters": schema.MapAttribute{
+				Description: "Tag Filters are used to filter by tags return a more specific list of results. If you specify multiple filters, the results returned will have only resources that match all the specified filters.",
```
[the remark this answers is not in the document]
Comment (unasra) [reply internal/service/ipam/api_ipam_host_data_source.go]: Fixed

File: internal/service/ipam/api_ipam_host_resource_test.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+	"github.com/infobloxopen/bloxone-go-client/ipam"
+	"github.com/infobloxopen/terraform-provider-bloxone/internal/acctest"
+)
+
+//TODO:add tests
+// The following require additional resource/data source objects to be supported.
+// - auto_generate_records
+// - addresses
+// - host_namesFi
```
[the remark this answers is not in the document]
Comment (unasra) [reply internal/service/ipam/api_ipam_host_resource_test.go]: Deleted

File: internal/service/ipam/api_ipam_host_resource_test.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ProtoV6ProviderFactories: acctest.ProtoV6ProviderFactories,
+		Steps: []resource.TestStep{
+			// Create and Read
+			{
+				Config: testAccIpamHostComment(name, "test comment"),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "comment", "test comment"),
```
[the remark this answers is not in the document]
Comment (unasra) [reply internal/service/ipam/api_ipam_host_resource_test.go:87]: Added

File: internal/service/ipam/model_ipamsvc_host_address.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+func (m *IpamsvcHostAddressModel) Expand(ctx context.Context, diags *diag.Diagnostics) *ipam.IpamsvcHostAddress {
+	if m == nil {
+		return nil
+	}
+	to := &ipam.IpamsvcHostAddress{
+		Address: m.Address.ValueString(),
+		Ref:     m.Ref.ValueString(),
+		Space:   m.Space.ValueString(),
```
[the remark this answers is not in the document]
Comment (unasra) [reply internal/service/ipam/model_ipamsvc_host_address.go]: Addressed

File: internal/service/ipam/model_ipamsvc_host_name.go (hunk 2 above)
Comment (unasra) [reply internal/service/ipam/model_ipamsvc_host_name.go]: Addressed

File: internal/service/ipam/model_ipamsvc_ipam_host.go (hunk 3 above)
Comment (unasra) [reply internal/service/ipam/model_ipamsvc_ipam_host.go]: Addressed

File: internal/service/ipam/model_ipamsvc_ipam_host.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mathewab) [reply internal/service/ipam/model_ipamsvc_ipam_host.go]: You should have kept optional. And added the default value of `false`. Now the field is marked read only, which is incorrect

[the remark this answers is not in the document]
Comment (unasra) [reply internal/service/ipam/model_ipamsvc_ipam_host.go:55]: fixed

**Review by mathewab — APPROVED** [review]

**Pull request merged**